src/pages/signin/SignInPage.js

```diff
@@ -0,0 +1,158 @@
+        login: PropTypes.string,
+        githubUsername: PropTypes.string,
+        password: PropTypes.string,
+        twoFactorAuthCode: PropTypes.string,
+    }),
+
+    // The session of the logged in person
+    session: PropTypes.shape({
+        // Error to display when there is a session error returned
+        error: PropTypes.string,
+    }),
+};
+
+const defaultProps = {
+    account: {},
+    session: null,
+    credentials: {},
+};
+
+class App extends Component {
+    componentDidMount() {
+        // Always reset the unread counter to zero on this page
+        updateUnread(0);
+    }
+
+    render() {
+        const session = this.props.session || {};
+
+        // If we end up on the sign in page and have an authToken then
+        // we are signed in and should be brought back to the site root
+        if (session.authToken) {
+            return <Redirect to={ROUTES.ROOT} />;
+        }
+
+        // @TODO figure out the real logic that needs to be used for showing the proper form
+        // It will depend on if they had an account already, had a Github username already,
+        // if they have access to this application, if they need to validate their account, etc.
+
+        // Show the login form if
+        // - A login has not been entered yet
+        const showLoginForm = !this.props.credentials.login;
+
+        // Show the GitHub username form if
+        // - A login has been entered
+        // - AND they do not have access to this app yet
+        // - AND the user hasn't entered a GitHub username yet
+        // - AND a password hasn't been entered yet
+        const showGithubUsernameForm = this.props.credentials.login
+            && !this.props.account.canAccessExpensifyCash
+            && !this.props.credentials.githubUsername
+            && !this.props.credentials.password;
+
+        // Show the password form if
+        // - A login has been entered
+        // - AND a GitHub username has been entered OR they already have access to expensify cash
+        // - AND an account exists already for this login
+        // - AND a password hasn't been entered yet
+        const showPasswordForm = this.props.credentials.login
+            && (
+                this.props.credentials.githubUsername
+                || this.props.account.canAccessExpensifyCash
+            )
+            && this.props.account.accountExists
+            && !this.props.credentials.password;
+
+        // Show the resend validation link form if
+        // - A login has been entered
+        // - AND a GitHub username has been entered OR they already have access to this app
+        // - AND an account did not exist for that login
+        const showResendValidationLinkForm = this.props.credentials.login
+            && (
+                this.props.credentials.githubUsername
+                || this.props.account.canAccessExpensifyCash
+            )
+            && !this.props.account.accountExists;
```

Comment:
NAB, this stuff could maybe be abstracted into a helper that returns the correct form to display and then we do or something similar...

```js
viewName === CONST.SIGN_IN.LOGIN && <LoginForm/>
``` 
Another question I have whether we could use routes like `/login` or `/github-name`? Rather than show a different view depending on the conditions of these various props it seems like we could just explicitly redirect to these views when some conditions are met or following the completion of some other action.

More or less just food for thought though. I think this is OK for now.

Reply:
That's an interesting thought about using routers. That could be a good way to organize this, but I'll save that for later. A possible downside is that each view would essentially need to control when it can/can't be displayed (in case someone uses the browser navigation to go between steps). That would make it a little bit more complex in the end I think.

Reply:
>A possible downside is that each view would essentially need to control when it can/can't be displayed

Not really sure what you mean maybe you can explain it another way? Views don't control when they can or can't be displayed you just display them or you do not display them based on a `Route` component.

Reply:
I mean that this logic that you commented on in the PR still needs to exist somewhere. It can't be entirely controlled by the Router. I'll give an example:

- A route for `/loginForm` is added to display `LoginForm`
- The `LoginForm` can only be displayed when `!credentials.login`
- A user enters their email into `LoginForm` and it's saved to `credentials.login`
- If a user manually navigates to `/loginForm` (by typing into the address bar, or using the back buttons on the browser), then there has to be some logic (I assume this would live inside the view) that would say "The router says I should show the `LoginForm`, but the user has already entered their login, so I should probably not show the `LoginForm` again.

Reply:
Ah ok thanks for the explanation! That's interesting. And I get what you are saying now. 

So I don't think we have to figure this out immediately. But this might come up again. Questions like...

- When should we create a new route for a view? (Probably, if it needs to be "deep linked" to or should respond to native navigation actions like back button etc)
- When we are _not_ creating a new route for a view or series of views - what does that look like? Is there some kind of best practice we can use to create a flow that has a series of steps. Does it bypass the usual navigation? What is it called? I think this is maybe a kind of "Stepper" navigation, but without feedback.

Reply:
> When should we create a new route for a view?

I think the criteria you started with is great.

> When we are not creating a new route for a view or series of views - what does that look like?

Yeah, it would be nice to make this consistent, though I think there is a risk of over-engineering this since we only have one use-case right now, and we don't have a clear idea of what other use cases might be. This could be something that we keep in mind for a refactoring the next time we notice a view has this multi-step approach to it.